### PR TITLE
pci-serial additions

### PR DESCRIPTION
**Pull request by ashwing123**
Added changes necessary to include `pci-serial` device into qcli as the console was not working with qemu-system-aarch64

### CODE REVIEW

File: chardevice.go [hunk 1]
```diff
@@ -170,10 +170,18 @@ func (cdev CharDevice) QemuParams(config *Config) []string {
...
-	if cdev.Driver != LegacySerial {
+	if cdev.Driver != LegacySerial && cdev.Driver != PCISerial {
 		qemuParams = append(qemuParams, "-device")
 		qemuParams = append(qemuParams, strings.Join(deviceParams, ","))
 	}
+	
+	//appending -serial none and -monitor none to qemuparams if we are using pci-serial
+	if cdev.Driver == PCISerial {
+		qemuParams = append(qemuParams, "-serial")
+		qemuParams = append(qemuParams, "none")
```
Comment (raharper) [inline chardevice.go:181]: I think this _may_ be required on qemu-system-aarch64, we can keep this for now;  Though I think we may not know whether an image we boot uses console=ttyS0 or console=ttyAMA0.  Need to think on this.

File: chardevice.go [hunk 2]
```diff
@@ -170,10 +170,18 @@ func (cdev CharDevice) QemuParams(config *Config) []string {
...
+		qemuParams = append(qemuParams, "-monitor")
+		qemuParams = append(qemuParams, "none")
```
Comment (raharper) [inline chardevice.go]: This is unrelated to the serial console, so drop this;  we'll continue to configure monitor as we do on x86.

File: qemu.go [hunk 3]
```diff
@@ -243,6 +243,7 @@ type Config struct {
 	CharDevices           []CharDevice           `yaml:"char-devices"`
 	LegacySerialDevices   []LegacySerialDevice   `yaml:"legacy-serial-devices"`
 	SerialDevices         []SerialDevice         `yaml:"serial-devices"`
+	PCISerialDevices 	  []PCISerialDevice	     `yaml:"pci-serial-devices"`
```
Comment (raharper) [inline qemu.go]: We don't need this, since the SerialDevice already contains a DeviceDriver object, so we can create a SerialDevice which uses Driver "pci-serial".

Comment (raharper) [reply qemu.go]: Take a look at serialdevice_test.go for the VirtioSerial case;  I think the PCISerial should look very similar to that (ignoring the extra options that a VirtioSerial device has)

```
$ qemu-system-x86_64 --device pci-serial,?
pci-serial options:
  x-pcie-lnksta-dllla=<bool> - on/off
  failover_pair_id=<str>
  chardev=<str>          - ID of a chardev to use as a backend
  romfile=<str>
  rombar=<uint32>
  addr=<int32>           - Slot and optional function number, example: 06.0 or 06
  prog_if=<uint8>
  x-pcie-extcap-init=<bool> - on/off
  command_serr_enable=<bool> - on/off
  multifunction=<bool>   - on/off
```

**Review by raharper — CHANGES_REQUESTED** [review]

File: qemu.go (hunk 3 above)
Comment (ashwing123) [reply qemu.go]: Ah yep sorry, looks pretty similar to VirtioSerial – only thing is I might need to add a chardev ID

File: chardevice_test.go [hunk 4]
```diff
@@ -7,6 +7,7 @@ var (
 	deviceCharDeviceBackendSocket   = "-chardev socket,id=serial0,path=/tmp/console.sock,server=on,wait=off"
 	deviceCharDeviceBackendStdioMux = "-chardev stdio,id=serial0,mux=on,signal=off"
 	deviceCharDeviceMultiple        = "-chardev socket,id=serial0,path=/tmp/console.sock,server=on,wait=off -chardev socket,id=monitor0,path=/tmp/monitor.sock,server=on,wait=off"
+	deviceCharDevicePCIDriver		= "-serial none -chardev socket,id=serial0,path=/tmp/console.sock,server=on,wait=off -device pci-serial,id=pciser0,chardev=serial0"
```
Comment (raharper) [inline chardevice_test.go]:
```suggestion
	deviceCharDevicePCIDriver	= "-serial none -chardev socket,id=serial0,path=/tmp/console.sock,server=on,wait=off -device pci-serial,id=pciser0,chardev=serial0"
```
non-printing space seems off

File: chardevice.go (hunk 1 above)
Comment (raharper) [reply chardevice.go:181]: OK, this isn't your fault -- with a few months off  from looking at this code I'm currently bothered by how in serialdevice_test.go, I have mostly Chardevices, and in chardev_test, all of them are using LegacySerial so it doesnt' emit a -device; but chardevs don't have devices at all, they have backends, which I've already modeled.

So, serialdevice and chardevice will need a refactor but let's merge this stuff first.

File: chardevice.go [hunk 5]
```diff
@@ -170,10 +170,16 @@ func (cdev CharDevice) QemuParams(config *Config) []string {
 	}
 
 	// Legacy serial is special. It does not follow the device + driver model
-	if cdev.Driver != LegacySerial {
+	if cdev.Driver != LegacySerial && cdev.Driver != PCISerial {
```
Comment (raharper) [inline chardevice.go]: Yeah, I don't like extending this, but will keep it for now.

In serialdevice_test.go, the VirtioSerialPortString test uses a CharDev object to emit the -device virtseiralport, but really, we should construct two objects:

```
chardev := CharDevice{
    // Driver:   VirtioSerialPort, // omit this here
    Backend:  Socket,
    ID:       "char0",
    Path:     "/tmp/char.sock",
}

serial := SerialDevice{
    Driver:    VirtioSerialPort,
    ID:        "channel0"
    CharDevID: "char0",
    Name:      "channel.0",
}

c := &Config{
    CharDevices:   []CharDevice{chardev},
    SerialDevices: []SerialDevice{serial},
   }
```

File: device.go [hunk 6]
```diff
@@ -166,6 +166,9 @@ const (
 
 	// TPM-CRB TPM Device
 	TPMCRBDebice DeviceDriver = "tpm-crb"
+
+	//PCI Serial Device
```
Comment (raharper) [inline device.go]:
```suggestion
	// PCI Serial Device
```

File: serialdevice.go [hunk 7]
```diff
@@ -102,6 +102,15 @@ func (dev LegacySerialDevice) deviceName(config *Config) string {
...
 
+//Virtio-serial specific driver struct
+type VirtioSerialDevice struct {
+	
+}
+
+//PCI-serial specific driver struct
+type PCISerialDevice struct {
+}
+
```
Comment (raharper) [inline serialdevice.go]:
```suggestion
```

We're dropping this stuff right?

File: serialdevice.go [hunk 8]
```diff
@@ -110,12 +119,16 @@ type SerialDevice struct {
...
 
-	// DisableModern prevents qemu from relying on fast MMIO.
-	DisableModern bool
+	// PCI Slot
+	Addr string
 
 	// ROMFile specifies the ROM file being used for this device.
 	ROMFile string
 
+	//virtio-serial specific attributes
```
Comment (raharper) [inline serialdevice.go:125]:
```suggestion
	// virtio-serial specific attributes
```

File: serialdevice.go [hunk 9]
```diff
@@ -124,6 +137,12 @@ type SerialDevice struct {
 
 	// MaxPorts is the maximum number of ports for this device.
 	MaxPorts uint
+
+	//pci-serial specific attributes
+	//Chardev associated with PCISerial
```
Comment (raharper) [inline serialdevice.go]:
```suggestion
	// pci-serial specific attributes
	// Chardev associated with PCISerial
```

File: serialdevice.go [hunk 10]
```diff
@@ -144,22 +167,34 @@ func (dev SerialDevice) QemuParams(config *Config) []string {
...
+		}
+		if dev.Transport.isVirtioCCW(config) {
+			if config.Knobs.IOMMUPlatform {
+				deviceParams = append(deviceParams, "iommu_platform=on")
+			}
+			deviceParams = append(deviceParams, fmt.Sprintf("devno=%s", dev.DevNo))
+		}
+	case PCISerial:
+		deviceParams = append(deviceParams, fmt.Sprintf("chardev=%s", dev.ChardevID))
+		if dev.Multifunction {
```
Comment (raharper) [inline serialdevice.go]: It's always worth checking the features of a given driver, VirtioSerial driver string is 'virtio-serial'; to see the options it supports you can run:

```
$ qemu-system-x86_64 -device virtio-serial,?
virtio-serial-pci options:
notify_on_empty=<bool> - on/off
use-started=<bool>
emergency-write=<bool> - on/off
max_ports=<uint32>
x-ignore-backend-features=<bool>
virtio-backend=<child<virtio-serial-device>>
packed=<bool>          - on/off
multifunction=<bool>   - on/off
page-per-vq=<bool>     - on/off
vectors=<uint32>
x-pcie-deverr-init=<bool> - on/off
disable-legacy=<OnOffAuto> - on/off/auto
x-pcie-pm-init=<bool>  - on/off
iommu_platform=<bool>  - on/off
ioeventfd=<bool>       - on/off
any_layout=<bool>      - on/off
failover_pair_id=<str>
ats=<bool>             - on/off
x-pcie-lnksta-dllla=<bool> - on/off
x-disable-pcie=<bool>  - on/off
x-pcie-flr-init=<bool> - on/off
addr=<int32>           - Slot and optional function number, example: 06.0 or 06
x-pcie-extcap-init=<bool> - on/off
x-pcie-lnkctl-init=<bool> - on/off
romfile=<str>
rombar=<uint32>
event_idx=<bool>       - on/off
disable-modern=<bool>
class=<uint32>
migrate-extra=<bool>   - on/off
virtio-pci-bus-master-bug-migration=<bool> - on/off
modern-pio-notify=<bool> - on/off
indirect_desc=<bool>   - on/off
command_serr_enable=<bool> - on/off
```

We're not implementing all of them, but you can see it does support the
multifunction parameter, let's move that up into the common path.

**Review by raharper — CHANGES_REQUESTED** [review]

File: serialdevice.go (hunk 10 above)
Comment (ashwing123) [reply serialdevice.go]: Sorry yeah completely missed this. Was able to check using `qemu-system-aarch64 --device virtio-serial,help` for aarch64

File: chardevice.go (hunk 1 above)
Comment (ashwing123) [reply chardevice.go:181]: Yeah I noticed that too. I can start drafting up changes for this after we get it merged

File: serialdevice.go (hunk 7 above)
Comment (ashwing123) [reply serialdevice.go]: Yep sorry

File: serialdevice.go [hunk 11]
```diff
@@ -134,7 +144,19 @@ func (dev SerialDevice) Valid() error {
...
-
+	if (dev.Driver == PCISerial) {
+		if (len(dev.ChardevIDs) > 4 || len(dev.ChardevIDs) == 0) {
+			return fmt.Errorf("PCISerialDevice has a malformed list of ChardevIDs (length 0 or length > 4)")
+		} 
+		if dev.ChardevIDs[0] == "" {
+			return fmt.Errorf("PCISerialDevice has no associated ChardevID")
+		}
+		if dev.MaxPorts != 1 && dev.MaxPorts != 2 && dev.MaxPorts != 4 {
+			return fmt.Errorf("PCISerialDevice has Ports not equal to 1, 2, or 4")
```
Comment (raharper) [inline serialdevice.go]:
```suggestion
			return fmt.Errorf("PCISerialDevice has MaxPorts not equal to 1, 2, or 4")
```

Comment (raharper) [reply serialdevice.go]: This seems reasonable.  pci-serial can run with out the chardev value; the serial port still works w.r.t the guestOS, one just can't connect that to qemu backend (like a file or stdio, etc).

But I don't think we have to provide that level of config at this time.

File: serialdevice.go [hunk 12]
```diff
@@ -134,7 +144,19 @@ func (dev SerialDevice) Valid() error {
 	if dev.ID == "" {
 		return fmt.Errorf("SerialDevice has empty ID field")
 	}
-
+	if (dev.Driver == PCISerial) {
+		if (len(dev.ChardevIDs) > 4 || len(dev.ChardevIDs) == 0) {
+			return fmt.Errorf("PCISerialDevice has a malformed list of ChardevIDs (length 0 or length > 4)")
```
Comment (raharper) [inline serialdevice.go]: I'd like to refer to this as `SerialDevice with Driver=PCISerialDevice`  I think that maps closer to what caller constructed:

```
dev := SerialDevice{
   Driver: PCISerialDevice,
   MaxPorts: 2,
}
```
seeing `SerialDevice` and `Driver=PCISerialDevice` should help caller find
where they're constructing that particular device.

**Review by raharper — CHANGES_REQUESTED** [review]

**Comment by raharper** [comment]
This PR was moved (and merged) here: https://github.com/project-machine/qcli/commit/41a9bce30c3bfae221378f621095126d017ac587
